## Equality
To check if two Python values, or variables, are equal you can use ==. To check for inequality, you need !=. As a refresher, have a look at the following examples that all result in True. Feel free to try them out in the IPython Shell.
````python
2 == (1 + 1)
"intermediate" != "python"
True != False
"Python" != "python"
````
When you write these comparisons in a script, you will need to wrap a print() function around them to see the output.

**Instructions**

- In the editor on the right, write code to see if True equals False.
- Write Python code to check if -5 * 15 is not equal to 75.
- Ask Python whether the strings "pyscript" and "PyScript" are equal.
- What happens if you compare booleans and integers? Write code to see if True and 1 are equal.

In [2]:
# Comparison of booleans

True == False

# Comparison of integers

-5 * 15 != 75

# Comparison of strings

"pyscript" == "PyScript"

# Compare a boolean with an integer

True == 1

True

## Greater and less than
In the video, Hugo also talked about the less than and greater than signs, < and > in Python. You can combine them with an equals sign: <= and >=. Pay attention: <= is valid syntax, but =< is not.

All Python expressions in the following code chunk evaluate to True:

`````python
3 < 4
3 <= 4
"alpha" <= "beta"
````
Remember that for string comparison, Python determines the relationship based on alphabetical order.

**Instructions**

- Write Python expressions, wrapped in a print() function, to check whether:
- x is greater than or equal to -10. x has already been defined for you.
- "test" is less than or equal to y. y has already been defined for you.
- True is greater than False.

In [3]:
# Comparison of integers
x = -3 * 6

print(x >= -10)

# Comparison of strings
y = "test"

print("test"<=y)

# Comparison of booleans

True > False

False
True


True

## Compare arrays
Out of the box, you can also use comparison operators with NumPy arrays.

Remember areas, the list of area measurements for different rooms in your house from Introduction to Python? This time there's two NumPy arrays: my_house and your_house. They both contain the areas for the kitchen, living room, bedroom and bathroom in the same order, so you can compare them.

**Instructions**

Using comparison operators, generate boolean arrays that answer the following questions:

- Which areas in my_house are greater than or equal to 18?
- You can also compare two NumPy arrays element-wise. Which areas in my_house are smaller than the ones in your_house?
- Make sure to wrap both commands in a print() statement so that you can inspect the output!

In [4]:
# Create arrays
import numpy as np
my_house = np.array([18.0, 20.0, 10.75, 9.50])
your_house = np.array([14.0, 24.0, 14.25, 9.0])

# my_house greater than or equal to 18

print(my_house >= 18)

# my_house less than your_house

print(my_house < your_house)

[ True  True False False]
[False  True  True False]


## and, or, not (1)
A boolean is either 1 or 0, True or False. With boolean operators such as and, or and not, you can combine these booleans to perform more advanced queries on your data.

In the sample code, two variables are defined: my_kitchen and your_kitchen, representing areas.

**Instructions**

- Write Python expressions, wrapped in a print() function, to check whether:
- my_kitchen is bigger than 10 and smaller than 18.
- my_kitchen is smaller than 14 or bigger than 17.
- double the area of my_kitchen is smaller than triple the area of your_kitchen.

In [5]:
# Define variables
my_kitchen = 18.0
your_kitchen = 14.0

# my_kitchen bigger than 10 and smaller than 18?

print(my_kitchen > 10 and my_kitchen < 18)

# my_kitchen smaller than 14 or bigger than 17?

print(my_kitchen < 14 or my_kitchen > 17)

# Double my_kitchen smaller than triple your_kitchen?

print(2 * my_kitchen < 3 * your_kitchen)

False
True
True


## and, or, not (2)
To see if you completely understood the boolean operators, have a look at the following piece of Python code:

x = 8
y = 9
not(not(x < 3) and not(y > 14 or y > 10))
What will the result be if you execute these three commands in the IPython Shell?

NB: Notice that not has a higher priority than and and or, it is executed first.

In [6]:
x = 8
y = 9
not(not(x < 3) and not(y > 14 or y > 10))

False

## Boolean operators with NumPy
Before, the operational operators like < and >= worked with NumPy arrays out of the box. Unfortunately, this is not true for the boolean operators and, or, and not.

To use these operators with NumPy, you will need np.logical_and(), np.logical_or() and np.logical_not(). Here's an example on the my_house and your_house arrays from before to give you an idea:
````python
np.logical_and(my_house > 13, 
               your_house < 15)
````
**Instructions**

- Generate boolean arrays that answer the following questions:
- Which areas in my_house are greater than 18.5 or smaller than 10?
- Which areas are smaller than 11 in both my_house and your_house? Make sure to wrap both commands in print() statement, so that you can inspect the output.

In [7]:
# Create arrays
import numpy as np
my_house = np.array([18.0, 20.0, 10.75, 9.50])
your_house = np.array([14.0, 24.0, 14.25, 9.0])

# my_house greater than 18.5 or smaller than 10

print(np.logical_or(my_house > 18.5, my_house < 10))

# Both my_house and your_house smaller than 11

print(np.logical_and(my_house < 11, your_house < 11))

[False  True False  True]
[False False False  True]


## if
It's time to take a closer look around in your house.

Two variables are defined in the sample code: room, a string that tells you which room of the house we're looking at, and area, the area of that room.

**Instructions**

- Examine the if statement that prints out "looking around in the kitchen." if room equals "kit".
- Write another if statement that prints out "big place!" if area is greater than 15.

In [8]:
# Define variables
room = "kit"
area = 14.0

# if statement for room
if room == "kit" :
    print("looking around in the kitchen.")

# if statement for area

if area > 15 :
    print("big place!")
    


looking around in the kitchen.


## Add else
In the script, the if construct for room has been extended with an else statement so that "looking around elsewhere." is printed if the condition room == "kit" evaluates to False.

Can you do a similar thing to add more functionality to the if construct for area?

**Instructions**

Add an else statement to the second control structure so that "pretty small." is printed out if area > 15 evaluates to False.

In [9]:
# Define variables
room = "kit"
area = 14.0

# if-else construct for room
if room == "kit" :
    print("looking around in the kitchen.")
else :
    print("looking around elsewhere.")

# if-else construct for area
if area > 15 :
    print("big place!")

else :
    print("pretty small.")

looking around in the kitchen.
pretty small.


## Customize further: elif
It's also possible to have a look around in the bedroom. The sample code contains an elif part that checks if room equals "bed". In that case, "looking around in the bedroom." is printed out.

It's up to you now! Make a similar addition to the second control structure to further customize the messages for different values of area.

**Instructions**

Add an elif to the second control structure such that "medium size, nice!" is printed out if area is greater than 10.

In [10]:
# Define variables
room = "bed"
area = 14.0

# if-elif-else construct for room
if room == "kit" :
    print("looking around in the kitchen.")
elif room == "bed":
    print("looking around in the bedroom.")
else :
    print("looking around elsewhere.")

# if-elif-else construct for area
if area > 15 :
    print("big place!")
elif area > 10 :
    print("medium size, nice!")
else :
    print("pretty small.")

looking around in the bedroom.
medium size, nice!


## Driving right (1)
Remember that cars dataset, containing the cars per 1000 people (cars_per_cap) and whether people drive right (drives_right) for different countries (country)? The code that imports this data in CSV format into Python as a DataFrame is included in the script.

In the video, you saw a step-by-step approach to filter observations from a DataFrame based on boolean arrays. Let's start simple and try to find all observations in cars where drives_right is True.

drives_right is a boolean column, so you'll have to extract it as a Series and then use this boolean Series to select observations from cars.

**Instructions**

- Extract the drives_right column as a Pandas Series and store it as dr.
- Use dr, a boolean Series, to subset the cars DataFrame. Store the resulting selection in sel.
- Print sel, and assert that drives_right is True for all observations.

In [2]:
# Import cars data
import pandas as pd
cars = pd.read_csv('cars.csv', index_col = 0)

# Extract drives_right column as Series: dr

dr = cars['drives_right']

# Use dr to subset cars: sel

sel = cars [dr]

# Print sel

print(sel)

     cars_per_cap        country  drives_right
US            809  United States          True
RU            200         Russia          True
MOR            70        Morocco          True
EG             45          Egypt          True


In [2]:
print(cars.columns)

Index([''cars_per_cap'', ''country'', ''drives_right''], dtype='object')


In [3]:
# Import cars data
import pandas as pd

# Read the cars data from CSV file
cars = pd.read_csv('cars.csv', index_col=0)

# Remove square brackets from column names
cars.columns = cars.columns.str.strip("[]")

# Extract drives_right column as Series: dr
dr = cars['drives_right']

# Use dr to subset cars: sel
sel = cars[dr]

# Print sel
print(sel)

KeyError: 'drives_right'

In [2]:
print(cars.columns)

Index([''cars_per_cap'', ''country'', ''drives_right''], dtype='object')


In [3]:
# Import cars data
import pandas as pd

# Read the cars data from CSV file
cars = pd.read_csv('cars.csv', index_col=0)

# Remove square brackets and quotation marks from column names
cars.columns = cars.columns.str.replace(r"['\"]", "")

# Extract drives_right column as Series: dr
dr = cars['drives_right']

# Use dr to subset cars: sel
sel = cars[dr]

# Print sel
print(sel)

C:\Users\Ornab\AppData\Local\Temp/ipykernel_5344/2867907475.py:8: FutureWarning: The default value of regex will change from True to False in a future version.
  cars.columns = cars.columns.str.replace(r"['\"]", "")


KeyError: 'drives_right'

In [6]:
print(babies.columns)

Index(['id', 'pluralty', 'outcome', 'date', 'gestation', 'sex', 'wt', 'parity',
       'race', 'age', 'ed', 'ht', 'wt1', 'drace', 'dage', 'ded', 'dht', 'dwt',
       'marital', 'inc', 'smoke', 'time', 'number'],
      dtype='object')


In [9]:
# Import cars data
import pandas as pd

# Read the cars data from CSV file
babies = pd.read_csv('babies.csv', index_col=0)

# Extract drives_right column as Series: dr
sex = babies['sex']

# Use dr to subset cars: sel
sel = babies[sex]

# Print sel
print(sel)

KeyError: "None of [Int64Index([1, 1, 1, 1, 1, 1, 1, 1, 1, 1,\n            ...\n            1, 1, 1, 1, 1, 1, 1, 1, 1, 1],\n           dtype='int64', length=1236)] are in the [columns]"

In [12]:
import pandas as pd

# Read the babies data from CSV file
babies = pd.read_csv('babies.csv')

# Extract sex column as Series: sex
sex = babies['sex']

# Use sex to subset babies: sel
sel = babies[sex == 1]

#make it one-liner

#sel = babies[babies['sex'] == 1]

# Print sel
print(sel)


      Unnamed: 0    id  pluralty  outcome  date  gestation  sex     wt  \
0              1    15         5        1  1411        284    1  120.0   
1              2    20         5        1  1499        282    1  113.0   
2              3    58         5        1  1576        279    1  128.0   
3              4    61         5        1  1504        999    1  123.0   
4              5    72         5        1  1425        282    1  108.0   
...          ...   ...       ...      ...   ...        ...  ...    ...   
1231        1232  9153         5        1  1672        275    1  113.0   
1232        1233  9163         5        1  1712        265    1  128.0   
1233        1234  9213         5        1  1672        291    1  130.0   
1234        1235  9229         5        1  1680        281    1  125.0   
1235        1236  9263         5        1  1668        297    1  117.0   

      parity  race  ...  drace  dage  ded  dht  dwt  marital  inc  smoke  \
0          1     8  ...      8    3

## Cars per capita (1)
Let's stick to the cars data some more. This time you want to find out which countries have a high cars per capita figure. In other words, in which countries do many people have a car, or maybe multiple cars.

Similar to the previous example, you'll want to build up a boolean Series, that you can then use to subset the cars DataFrame to select certain observations. If you want to do this in a one-liner, that's perfectly fine!

**Instructions**

- Select the cars_per_cap column from cars as a Pandas Series and store it as cpc.
- Use cpc in combination with a comparison operator and 500. You want to end up with a boolean Series that's True if the corresponding country has a cars_per_cap of more than 500 and False otherwise. Store this boolean Series as many_cars.
- Use many_cars to subset cars, similar to what you did before. Store the result as car_maniac.
- Print out car_maniac to see if you got it right.

In [1]:
# Import cars data
import pandas as pd
cars = pd.read_csv('cars.csv', index_col = 0)

# Create car_maniac: observations that have a cars_per_cap over 500

cpc = cars['cars_per_cap']

many_cars = cpc > 500

car_maniac = cars [many_cars]



# Print car_maniac

print(car_maniac)

     cars_per_cap        country  drives_right
US            809  United States          True
AUS           731      Australia         False
JAP           588          Japan         False


## Cars per capita (2)
Remember about np.logical_and(), np.logical_or() and np.logical_not(), the NumPy variants of the and, or and not operators? You can also use them on Pandas Series to do more advanced filtering operations.

Take this example that selects the observations that have a cars_per_cap between 10 and 80. Try out these lines of code step by step to see what's happening.
````python
cpc = cars['cars_per_cap']
between = np.logical_and(cpc > 10, cpc < 80)
medium = cars[between]
````
**Instructions**

- Use the code sample provided to create a DataFrame medium, that includes all the observations of cars that have a cars_per_cap between 100 and 500.
- Print out medium.

In [3]:
# Import cars data
import pandas as pd
cars = pd.read_csv('cars.csv', index_col = 0)

# Import numpy, you'll need this
import numpy as np

# Create medium: observations with cars_per_cap between 100 and 500

cpc = cars['cars_per_cap']

#medium = np.logical_and(cars['cars_per_cap'] > 100, cars['cars_per_cap'] < 500)

between = np.logical_and(cpc > 100, cpc < 500)

medium = cars[between]

# Print medium

print(medium)

    cars_per_cap country  drives_right
RU           200  Russia          True


## while: warming up
The while loop is like a repeated if statement. The code is executed over and over again, as long as the condition is True. Have another look at its recipe.

while condition :
    expression
Can you tell how many printouts the following while loop will do?
````python
x = 1
while x < 4 :
    print(x)
    x = x + 1
````

In [4]:
x = 1
while x < 4 :
    print(x)
    x = x + 1
    

1
2
3


## Basic while loop
Below you can find the example from the video where the error variable, initially equal to 50.0, is divided by 4 and printed out on every run:
````python
error = 50.0
while error > 1 :
    error = error / 4
    print(error)
````
This example will come in handy, because it's time to build a while loop yourself! We're going to code a while loop that implements a very basic control system for an inverted pendulum. If there's an offset from standing perfectly straight, the while loop will incrementally fix this offset.

Note that if your while loop takes too long to run, you might have made a mistake. In particular, remember to indent the contents of the loop using four spaces or auto-indentation!

**Instructions**

- Create the variable offset with an initial value of 8.
- Code a while loop that keeps running as long as offset is not equal to 0. Inside the while loop:
    - Print out the sentence "correcting...".
    - Next, decrease the value of offset by 1. You can do this with offset = offset - 1.
    - Finally, still within your loop, print out offset so you can see how it changes.

In [5]:
# Initialize offset

offset = 8

# Code the while loop

while offset != 0 :
    print("correcting...")
    offset = offset - 1
    print(offset)

correcting...
7
correcting...
6
correcting...
5
correcting...
4
correcting...
3
correcting...
2
correcting...
1
correcting...
0


## Add conditionals
The while loop that corrects the offset is a good start, but what if offset is negative? You can try to run the following code where offset is initialized to -6:
````python
#Initialize offset
offset = -6

#Code the while loop
while offset != 0 :
    print("correcting...")
    offset = offset - 1
    print(offset)
````
but your session will be disconnected. The while loop will never stop running, because offset will be further decreased on every run. offset != 0 will never become False and the while loop continues forever.

Fix things by putting an if-else statement inside the while loop. If your code is still taking too long to run, you probably made a mistake!

**Instructions**

- Inside the while loop, complete the if-else statement:
    - If offset is greater than zero, you should decrease offset by 1.
    - Else, you should increase offset by 1.
- If you've coded things correctly, hitting Submit Answer should work this time.

If your code is still taking too long to run (or your session is expiring), you probably made a mistake. Check your code and make sure that the statement offset != 0 will eventually evaluate to FALSE!

In [6]:
# Initialize offset
offset = -6

# Code the while loop
while offset != 0 :
    print("correcting...")
    if offset > 0 :
      offset = offset - 1
    else : 
      offset = offset + 1    
    print(offset)

correcting...
-5
correcting...
-4
correcting...
-3
correcting...
-2
correcting...
-1
correcting...
0


## Loop over a list
Have another look at the for loop that Hugo showed in the video:
````python
fam = [1.73, 1.68, 1.71, 1.89]
for height in fam : 
    print(height)
````
As usual, you simply have to indent the code with 4 spaces to tell Python which code should be executed in the for loop.

The areas variable, containing the area of different rooms in your house, is already defined.

**Instructions**

Write a for loop that iterates over all elements of the areas list and prints out every element separately.

In [7]:
# areas list
areas = [11.25, 18.0, 20.0, 10.75, 9.50]

# Code the for loop

for el in areas :
    print(el)

11.25
18.0
20.0
10.75
9.5


## Indexes and values (1)
Using a for loop to iterate over a list only gives you access to every list element in each run, one after the other. If you also want to access the index information, so where the list element you're iterating over is located, you can use enumerate().

As an example, have a look at how the for loop from the video was converted:
````python
fam = [1.73, 1.68, 1.71, 1.89]
for index, height in enumerate(fam) :
    print("person " + str(index) + ": " + str(height))
````
**Instructions**

- Adapt the for loop in the sample code to use enumerate() and use two iterator variables.
- Update the print() statement so that on each run, a line of the form "room x: y" should be printed, where x is the index of the list element and y is the actual list element, i.e. the area. Make sure to print out this exact string, with the correct spacing.

In [8]:
# areas list
areas = [11.25, 18.0, 20.0, 10.75, 9.50]

# # Change for loop to use enumerate() and update print()
# for a in areas :
#     print(a)

# Change for loop to use enumerate() and update print()
for index, area in enumerate(areas) :
    print("room " + str(index) + ": " + str(area))

room 0: 11.25
room 1: 18.0
room 2: 20.0
room 3: 10.75
room 4: 9.5


## Indexes and values (2)
For non-programmer folks, room 0: 11.25 is strange. Wouldn't it be better if the count started at 1?

**Instructions**

Adapt the print() function in the for loop so that the first printout becomes "room 1: 11.25", the second one "room 2: 18.0" and so on.

In [9]:
# areas list
areas = [11.25, 18.0, 20.0, 10.75, 9.50]

# Code the for loop

for index, area in enumerate(areas) :
    print("room " + str(index + 1) + ": " + str(area))

room 1: 11.25
room 2: 18.0
room 3: 20.0
room 4: 10.75
room 5: 9.5


## Loop over list of lists
Remember the house variable from the Intro to Python course? Have a look at its definition in the script. It's basically a list of lists, where each sublist contains the name and area of a room in your house.

It's up to you to build a for loop from scratch this time!

**Instructions**

Write a for loop that goes through each sublist of house and prints out the x is y sqm, where x is the name of the room and y is the area of the room.

In [11]:
# house list of lists
house = [["hallway", 11.25], 
         ["kitchen", 18.0], 
         ["living room", 20.0], 
         ["bedroom", 10.75], 
         ["bathroom", 9.50]]
         
# Build a for loop from scratch
for room in house :
    print("the " + room[0] + " is " + str(room[1]) + " sqm. ")

# for x, room in enumerate(house) :
#     # print("Room " + str(x+1) + ": The " + room[0] + " is " + str(room[1]) + " sqm.")
#     print("the " + room[0] + " is " + str(room[1]) + " sqm. ")


the hallway is 11.25 sqm. 
the kitchen is 18.0 sqm. 
the living room is 20.0 sqm. 
the bedroom is 10.75 sqm. 
the bathroom is 9.5 sqm. 


## Loop over dictionary
In Python 3, you need the items() method to loop over a dictionary:
````python
world = { "afghanistan":30.55, 
          "albania":2.77,
          "algeria":39.21 }

for key, value in world.items() :
    print(key + " -- " + str(value))
````
Remember the europe dictionary that contained the names of some European countries as key and their capitals as corresponding value? Go ahead and write a loop to iterate over it!

**Instructions**

Write a for loop that goes through each key:value pair of europe. On each iteration, "the capital of x is y" should be printed out, where x is the key and y is the value of the pair.

In [1]:
# Definition of dictionary
europe = {'spain':'madrid', 'france':'paris', 'germany':'berlin',
          'norway':'oslo', 'italy':'rome', 'poland':'warsaw', 'austria':'vienna' }
          
# Iterate over europe

for x, y in europe.items() :
    print("The capital of " + x + " is " + str(y))

The capital of spain is madrid
The capital of france is paris
The capital of germany is berlin
The capital of norway is oslo
The capital of italy is rome
The capital of poland is warsaw
The capital of austria is vienna


## Loop over NumPy array
If you're dealing with a 1D NumPy array, looping over all elements can be as simple as:
````python
for x in my_array :
    ...
````
If you're dealing with a 2D NumPy array, it's more complicated. A 2D array is built up of multiple 1D arrays. To explicitly iterate over all separate elements of a multi-dimensional array, you'll need this syntax:
````python
for x in np.nditer(my_array) :
    ...
````
Two NumPy arrays that you might recognize from the intro course are available in your Python session: np_height, a NumPy array containing the heights of Major League Baseball players, and np_baseball, a 2D NumPy array that contains both the heights (first column) and weights (second column) of those players.

**Instructions**

- Import the numpy package under the local alias np.
- Write a for loop that iterates over all elements in np_height and prints out "x inches" for each element, where x is the value in the array.
- Write a for loop that visits every element of the np_baseball array and prints it out.

In [4]:
# Import numpy as np

import numpy as np

np_height = np.array([74, 74, 72, 75, 75, 73])
np_baseball = np.array([[74, 180],
                        [74, 215],
                        [72, 210],
                        [75, 205],
                        [75, 190],
                        [73, 195]])



# For loop over np_height

for x in np_height :
    print (str(x) + " inches ")


# For loop over np_baseball
for x in np.nditer(np_baseball) :
    print (str(x))

74 inches 
74 inches 
72 inches 
75 inches 
75 inches 
73 inches 
74
180
74
215
72
210
75
205
75
190
73
195


## Loop over DataFrame (1)
Iterating over a Pandas DataFrame is typically done with the iterrows() method. Used in a for loop, every observation is iterated over and on every iteration the row label and actual row contents are available:
````python
for lab, row in brics.iterrows() :
    ...
````
In this and the following exercises you will be working on the cars DataFrame. It contains information on the cars per capita and whether people drive right or left for seven countries in the world.

**Instructions**

Write a for loop that iterates over the rows of cars and on each iteration perform two print() calls: one to print out the row label and one to print out all of the rows contents.

In [5]:
# Import cars data
import pandas as pd
cars = pd.read_csv('cars.csv', index_col = 0)

# Iterate over rows of cars

for lab, row in cars.iterrows() :
    print(lab)
    print(str(row))

US
cars_per_cap              809
country         United States
drives_right             True
Name: US, dtype: object
AUS
cars_per_cap          731
country         Australia
drives_right        False
Name: AUS, dtype: object
JAP
cars_per_cap      588
country         Japan
drives_right    False
Name: JAP, dtype: object
IN
cars_per_cap       18
country         India
drives_right    False
Name: IN, dtype: object
RU
cars_per_cap       200
country         Russia
drives_right      True
Name: RU, dtype: object
MOR
cars_per_cap         70
country         Morocco
drives_right       True
Name: MOR, dtype: object
EG
cars_per_cap       45
country         Egypt
drives_right     True
Name: EG, dtype: object


## Loop over DataFrame (2)
The row data that's generated by iterrows() on every run is a Pandas Series. This format is not very convenient to print out. Luckily, you can easily select variables from the Pandas Series using square brackets:
````python
for lab, row in brics.iterrows() :
    print(row['country'])
````
**Instructions**

- Using the iterators lab and row, adapt the code in the for loop such that the first iteration prints out "US: 809", the second iteration "AUS: 731", and so on.
- The output should be in the form "country: cars_per_cap". Make sure to print out this exact string (with the correct spacing).
    - You can use str() to convert your integer data to a string so that you can print it in conjunction with the country label.

In [6]:
# Import cars data
import pandas as pd
cars = pd.read_csv('cars.csv', index_col = 0)

# Adapt for loop
for lab, row in cars.iterrows() :
    print(lab + ": " + str(row['cars_per_cap']))

US: 809
AUS: 731
JAP: 588
IN: 18
RU: 200
MOR: 70
EG: 45


## Add column (1)
In the video, Hugo showed you how to add the length of the country names of the brics DataFrame in a new column:
````python
for lab, row in brics.iterrows() :
    brics.loc[lab, "name_length"] = len(row["country"])
````
You can do similar things on the cars DataFrame.

**Instructions**

- Use a for loop to add a new column, named COUNTRY, that contains a uppercase version of the country names in the "country" column. You can use the string method upper() for this.
- To see if your code worked, print out cars. Don't indent this code, so that it's not part of the for loop.

In [7]:
# Import cars data
import pandas as pd
cars = pd.read_csv('cars.csv', index_col = 0)

# Code for loop that adds COUNTRY column

for lab, row in cars.iterrows() :
    cars.loc[lab, 'COUNTRY'] = str.upper(row['country'])


# Print cars
print(cars)

     cars_per_cap        country  drives_right        COUNTRY
US            809  United States          True  UNITED STATES
AUS           731      Australia         False      AUSTRALIA
JAP           588          Japan         False          JAPAN
IN             18          India         False          INDIA
RU            200         Russia          True         RUSSIA
MOR            70        Morocco          True        MOROCCO
EG             45          Egypt          True          EGYPT


## Add column (2)
Using iterrows() to iterate over every observation of a Pandas DataFrame is easy to understand, but not very efficient. On every iteration, you're creating a new Pandas Series.

If you want to add a column to a DataFrame by calling a function on another column, the iterrows() method in combination with a for loop is not the preferred way to go. Instead, you'll want to use apply().

Compare the iterrows() version with the apply() version to get the same result in the brics DataFrame:
````python
for lab, row in brics.iterrows() :
    brics.loc[lab, "name_length"] = len(row["country"])

brics["name_length"] = brics["country"].apply(len)
````
We can do a similar thing to call the upper() method on every name in the country column. However, upper() is a method, so we'll need a slightly different approach:

**Instructions**

- Replace the for loop with a one-liner that uses .apply(str.upper). The call should give the same result: a column COUNTRY should be added to cars, containing an uppercase version of the country names.
- As usual, print out cars to see the fruits of your hard labor

In [8]:
# Import cars data
import pandas as pd
cars = pd.read_csv('cars.csv', index_col = 0)


# for lab, row in cars.iterrows() :
#     cars.loc[lab, "COUNTRY"] = row["country"].upper()

# Use .apply(str.upper)

cars["COUNTRY"] = cars["country"].apply(str.upper)

print(cars)

     cars_per_cap        country  drives_right        COUNTRY
US            809  United States          True  UNITED STATES
AUS           731      Australia         False      AUSTRALIA
JAP           588          Japan         False          JAPAN
IN             18          India         False          INDIA
RU            200         Russia          True         RUSSIA
MOR            70        Morocco          True        MOROCCO
EG             45          Egypt          True          EGYPT
